DOCSP-12158: Fundamentals Documents

ccho-mongodb opened the pull request:
## Pull Request Info

### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-12158

### Docs staging link (requires sign-in on MongoDB Corp SSO):
https://docs-mongodbcom-staging.corp.mongodb.com/b067e65/java/docsworker-xlarge/DOCSP-12158-fundamentals-documents/fundamentals/data-formats/documents

### Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Does it render on staging correctly?
- [x] Are all the links working?
- [x] Are the staging links in the PR description updated?

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
-- Link to Fundamentals > CRUD for more information
-- Link to API documentation
+The ``Document`` class offers a flexible representation of a BSON document.
+You can access and manipulate document data using Java types with this
```
```suggestion
You can access and manipulate fields using Java types with this
```
Maybe this is a good opportunity to use the "field" term to familiarize readers? (see previous comment)

ccho-mongodb replied:
Appending the fix to another commit.

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+     - ``java.lang.Long``
+   * - Null
+     - ``null``
+   * - ObjectId``
```
```suggestion
   * - ObjectId
```
I think you accidentally a monospace!

ccho-mongodb replied:
Whoops, good catch. Even the autobuilder did not warn me about that.

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+:doc:`Codecs </fundamentals/data-formats/codecs>`.
+
+In the following code snippet, we show how to instantiate and build a sample
+``Document`` instance that contains a variety of data types:
```
Thought: Is "contains a variety of data types" the best descriptor we can offer here? I wonder if this glosses over the fact that a document actually contains a collection of string keys that map to values with a variety of data types. Perhaps we can add some more detail about that fact here (and, my favorite bone to pick at this point, add the term "field" here)?

ccho-mongodb replied:
I think you might have been expecting a description of a Document class. I was actually attempting to communicate that the example intentionally uses field values of different types to demonstrate how to set them. I'll clarify that here. 

Regarding data structures contained within the Document class: I want to limit the scope of what is discussed in reference documentation since the reader can refer to API documentation for those details.

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
-- Link to API documentation
+The ``Document`` class offers a flexible representation of a BSON document.
+You can access and manipulate document data using Java types with this
+class. See the following table for mappings between frequently-used BSON
```
nit: should "frequently-used" be "frequently used"? I tried to look it up in the meta style guide but I couldn't find it there.

ccho-mongodb replied:
I think usually you use hyphen for adverb + past participle verb, but maybe open for debate.
There's some discussion here:
https://english.stackexchange.com/questions/225433/should-i-used-a-hyphen-for-often-used-and-well-documented

I'm going to leave it _as-is_ for now, unless there's a pressing need to change it.

nathan-contino-mongo replied:
Sounds good to me!

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+The ``BsonDocument`` class provides a type-safe API to access and manipulate
+a BSON document. You need to specify the BSON type from the Java BSON
+library for each field. See the following table for mappings between
+frequently-used BSON and Java BSON library types:
```
duplicate of an earlier nit: should "frequently-used" be "frequently used"? I tried to look it up in the meta style guide but I couldn't find it there.

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+     - ``org.bson.BsonString``
+
+
+You can also customize the type mapping by specifying a custom codec.
```
```suggestion
You can customize the type mapping by specifying a custom codec.
```
Nit: It seems to me like "also" is unnecessary here. What do you think?

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+   * - String
+     - ``java.lang.String``
+
+You can also customize the type mapping by specifying a custom codec.
```
Nit: It seems like "also" is unnecessary here. What do you think?
```suggestion
You can customize the type mapping by specifying a custom codec.
```

ccho-mongodb replied:
I was also waffling with the wording of this sentence. I think I might need to provide more detail as I was hoping I could keep it concise and that the reader could infer that the table showed default mappings and that they could override them by specifying or implementing a custom codec.

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+
+  // MongoClient mongoClient = ...
+
+  MongoDatabase database = mongoClient.getDatabase("test_data");
```
Because Document is the default representation for database documents, it seems like you're just relying on the defaults here. Should we call this out explicitly above?

nathan-contino-mongo replied:
The equivalent code snippets for other document representations use the "test_fundamentals" database instead of "test_data". Should we standardize the database used?

ccho-mongodb replied:
> Because Document is the default representation for database documents, it seems like you're just relying on the defaults here. Should we call this out explicitly above?

I'm guessing you meant this on the next line for `getCollection`. I'll add a mention and link to the API docs.

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+   // MongoClient mongoClient = ...
+
+   MongoCollection<BsonDocument> collection = database.getCollection("authors", BsonDocument.class);
+   MongoDatabase database = mongoClient.getDatabase("fundamentals_data");
```
I might be going crazy but I *think* these lines are in the wrong order!

ccho-mongodb replied:
Yeah, good copypasta catch.

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+You can insert this document into a collection using the following code:
+
+.. code-block:: java
+   :copyable: true
```
While you have specified that these blocks should be copyable, I don't see a copy button in your staged version. Nextgen bug?

ccho-mongodb replied:
Will investigate. When I last checked, the first code blocks I created and previewed were working fine, but maybe I introduced some syntax issue along the way

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+   import com.mongodb.client.model.Filters;
+   ...
+
+   BsonDocument doc = collection.find(Filters.eq("novels.yearPublished", 1967)).first();
```
I wonder if this query could confuse users who aren't hugely familiar with MongoDB, since we're using some fairly advanced ideas here (querying an embedded document with dot syntax, matching a document based off a match to one field in a single document in a collection). On one hand, this does demonstrate some interesting MongoDB edge cases, and a lot of readers might be curious about dot syntax. On the other hand... when I started out with MongoDB, I might have had some trouble following this. Perhaps it would be better to use a simpler filter here and instead cover these more advanced concepts in the CRUD section?

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+For more information on retrieving and manipulating MongoDB data, see our
+:doc:`CRUD guide </fundamentals/crud>`.
+
+For more information on this class, see the
```
Duplicate: perhaps a good use for a see also directive?

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+The ``BasicDBObject`` class allows you to access and manipulate document data
+using Java types. We recommend that you avoid using this class unless you
+are migrating an application from that uses an older driver version. In
+addition to deprecation, this class has the following limitations:
```
Instead of the mildly awkward "in addition to deprecation" phrasing, I wonder if we could:
1. Add (Deprecated) after the section header, akin to the (BETA) label for Realm Sync
2. Mention "deprecated" in the first sentence ("The deprecated ``BasicDBObject`` class") to make the deprecation even more obvious.

We could then just change this sentence to "This class has the following limitations" and, if we really want to hammer in the deprecation idea, add "This class is deprecated" as a new first bullet point in the limitations.

ccho-mongodb replied:
Thought it was deprecated, but I guess not.

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+addition to deprecation, this class has the following limitations:
+
+- This class implements the ``DBObject`` interface, so you cannot extend its
+  API without breaking binary compatibility.
```
Perhaps I'm just not familiar with "binary compatibility" or API extension constraints, but it's unclear to me how the first clause relates to the second clause (I'm assuming a causal relationship). It could also be helpful to define "binary compatibility" here (are we talking about binaries as in programs? Or binary representation of objects in process memory?

ccho-mongodb replied:
I copied this from the existing docs, but I'm assuming it is referring to https://wiki.eclipse.org/Evolving_Java-based_APIs_2#Evolving_API_packages -- see example 4. (A more recent doc: https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html )

As I understand it, "extend its API" means add a method to a class that implements it such as BasicDBObject. Suppose you had a compiled app that uses driver v4.1, but v4.2 came out with a new method, BasicDBObject#coolNewMethod. Normally, you could just upgrade the jar in the classpath and restart the app (or hot-reload) and it would work fine.  However, since you broke binary compatibility with v4.2 by adding a method, you need to recompile your app for it to work with v4.2. However, if there was some non-breaking change such as a private method change, your app would continue to work without recompilation because it's not visible to your binary.

Admittedly, it is a little unclear which object the pronoun "its" refers to so I can reword. It's also probably less frequently encountered than the next item, so I'll re-order it in the list.

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+
+You can insert this document into a collection using the following code:
+
+.. code-block:: java
```
For some reason, syntax highlighting doesn't seem to be working for this example.

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+               + ", yearPublished: " + novel.getInt("yearPublished"));
+           }
+       }
+   }
```
Similar thoughts as I had for the Document class -- should we explain the usage and constraints of the BasicDBObject type getters?

ccho-mongodb replied:
Addressed in a note directive

nathan-contino-mongo commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+  why you might prefer one over the other.
+- Provided usage examples for each class on building documents containing
+  multiple types, inserting them into a collection, and
+  retrieving/accessing their typed fields.
```
Nice summary, I really like this.

ccho-mongodb replied:
Thanks!

nathan-contino-mongo reviewed (commented):
I think this is a great start -- a lot of comments from me, but mostly suggestions of how we can enhance this guide to make it even better. This is a tricky concept to cover and I think we're definitely moving in the right direction!

ccho-mongodb commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+binary JSON (BSON) format and is the base unit in the MongoDB data hierarchy.
+
+Each **document** belongs to a single **collection** and each **collection**
+belongs to a single **database**. For more information, see our MongoDB manual
```
Thanks for bringing this up.

1) I think I could replace this with `:term:` (see Glossary terms in https://docs.mongodb.com/meta/style-guide/style/text-formatting/).
Italics could be used for introducing new terms (see "New terms" in https://docs.mongodb.com/meta/style-guide/style/text-formatting/ ) explained immediately afterwards

2) I did think about creating a graphic and searched through server/manual documentation and wasn't able to find anything suitable to link to or re-use. Maybe I can't find the right page which probably means our users can't find it either. Instead of attempting to explain it from scratch, I just wanted to introduce repetition of terms and what they are as a learning aid.

That said, if this was a language-specific concept, I would probably devote a section for it.

nathan-contino-mongo replied:
I think this is a perfect opportunity for creating a glossary, so definitely think about that! Just check to make sure it's supported properly by nextgen first :)

I hoped that the linked server page would include such a graphic, too, but sadly it doesn't. Might be a good idea to ask in the #docs chat to see if anybody knows of such a graphic (or would be willing to create one for you, if they have a LucidChart license and some degree of artistic skills)

nathan-contino-mongo replied:
+1 to repetition as a learning aid, I think that works quite well here.

ccho-mongodb replied:
Didn't realize you need a glossary to use `:term:`. I removed formatting for the time being. I'm a little confused by the inconsistent formatting here, so probably something to address sometime later: https://github.com/mongodb/docs-java/blob/master/source/fundamentals/databases-collections.txt

ccho-mongodb commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+   * - ``BasicDBObject``
+     - ``com.mongodb``
+     - No
+     - When you are migrating your application from a legacy driver version.
```
Thanks, I wonder if this should be a pattern we reuse for pages that present multiple similar topics.

ccho-mongodb commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+.. code-block:: java
+   :copyable: true
+
+  // MongoClient mongoClient = ...
```
It was designed to shows the type with the instantiation as the variable is used later on. I can certainly add more explanation.

nathan-contino-mongo replied:
I think that your solution is solid here -- it gives just enough of a hint that a reader can google for a solution (or search on the docs, or navigate within the docs) if they don't already know how to instantiate a mongoClient, mongoCollection, etc.

ccho-mongodb commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+   :copyable: true
+
+   import com.mongodb.client.model.Filters;
+   ...
```
From [this commen](https://github.com/mongodb/docs-java/pull/39#discussion_r527092414)t, I need to add some explanation of the dots.

ccho-mongodb commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+:doc:`CRUD guide </fundamentals/crud>`.
+
+For more information on this class, see the
+:java-docs:`Document API documentation </apidocs/bson/org/bson/Document.html>`.
```
Thanks for the suggestion. I think you mean `seealso`. I've been avoiding using it, but I think it doesn't really change the experience either way. I'll keep it in mind though since the change in formatting it produces could be a good signal to readers if used consistently.

ccho-mongodb commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
-- Recommend Document for general use and why.
-- Duplicate field name admonition (I forgot what this is)
+A MongoDB **document** is a data structure that contains key/value pairs in
+binary JSON (BSON) format and is the base unit in the MongoDB data hierarchy.
```
I'll replace this with something similar.

ccho-mongodb commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+.. warning:: Duplicate Field Names
+
+   Although a document is valid if it contains two duplicate keys, the driver
+   only sets the value of the key to the value in the last instance.
```
I don't have any qualms with adding this info, but "field" is included in the linked doc: https://docs.mongodb.com/manual/core/document/#field-names

I'll make sure I explain that the link has important info about field-related terminology and limitations, and remove this warning. I think the reason it is called out in the current version of the driver docs is because they don't link to the manual entry: https://mongodb.github.io/mongo-java-driver/4.1/bson/documents/

ccho-mongodb commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,351 @@ Documents
+           System.out.println("title: " + novel.getString("title")
+               + ", yearPublished: " + novel.getInteger("yearPublished"));
+           });
+   }
```
>  include a sentence or two clarifying how you access field values from an Instance of Document

Could you clarify what you mean by this?  I included methods called on the Document instance to access the fields such as getObjectId, getString, and getDate. I also showed how to access other types within an embedded Document.

For the second part:
>what happens if you use the wrong type of get (for instance, getString on an ObjectId field)

I'll think about it how I might incorporate that info.

nathan-contino-mongo replied:
I very much like the note that you added after these examples -- that's precisely what I was suggesting above. Resolved!

MongoCaleb commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,379 @@ Documents
+.. code-block:: java
+
+   Document author = new Document("_id", new ObjectId())
+       .append("name", "Gabriel García Márquez")
```
Nice author reference!

MongoCaleb commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,379 @@ Documents
+   type :java-docs:`BsonValue </apidocs/bson/org/bson/BsonValue.html>` and to
+   skip type checking.
+
+The output should resemble the following:
```
"will" instead of "should"?

MongoCaleb commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,379 @@ Documents
+   You can call the :java-docs:`get() </apidocs/bson/org/bson/BasicBSONObject.html?is-external=true#get(java.lang.String)>`
+   method to retrieve values as type type ``Object`` and to skip type checking.
+
+The output should resemble the following:
```
"will" instead of "should"?

ccho-mongodb replied:
I avoid using "will" across docs because it often leads to passive voice and I think we prefer active when possible based on some feedback I received last year. Do you think that I should change it anyway?

MongoCaleb commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,379 @@ Documents
+   method specified by the ``Map`` interface to retrieve field values as
+   type ``Object`` and to skip type checking.
+
+The output should resemble the following:
```
"will" instead of "should"?

MongoCaleb reviewed (approved):
LGTM with one suggestion that occurs 3x

ccho-mongodb commented:
Hi @jyemin and @rozza , ready for your review whenever you get the chance. Thanks!

nathan-contino-mongo reviewed (approved):
LGTM -- this page is really coming together! Ready for tech review.

rozza reviewed (approved):
LGTM.

I'm not sold on having DBObjects listed but I suppose people are still using them! :cry:

jyemin commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,379 @@ Documents
-- Duplicate field name admonition (I forgot what this is)
+A MongoDB **document** is a data structure that contains key/value pairs in
+binary JSON (BSON) format. The document is the base unit of the MongoDB
+data hierarchy: each document belongs to a single collection
```
> The document is the base unit of the MongoDB	data hierarchy

I feel like you're inventing new terminology here: I've never heard a document described this way in our documentation.  And it's not accurate.  The base unit is a BSON _value_ (e.g. int64, int32, boolean), not a BSON document.

And also, documents are used for other purposed besides storage in collections: e.g. commands, queries, sort criteria, etc

ccho-mongodb replied:
Good point, I'll remove the phrase that mentions it's the base unit.

jyemin commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,379 @@ Documents
-- Link to Fundamentals > CRUD for more information
-- Link to API documentation
+The ``Document`` class offers a flexible representation of a BSON document.
+You can access and manipulate fields using Java types with this class. See
```
"using Java types from the standard library"  would be more accurate.  BsonDocument also uses Java types, but ones that are defined in our own bson library

ccho-mongodb replied:
Thanks, will update.

jyemin commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,379 @@ Documents
+
+   Document author = new Document("_id", new ObjectId())
+       .append("name", "Gabriel García Márquez")
+       .append("lastUpdated", new Date(120, 2, 26))
```
That's a strange date... The year 120?  

Also, that constructor is deprecated.  Horribly, the way to do this in Java is:

    Date.from(LocalDate.of(2000, 1, 1).atStartOfDay(ZoneId.systemDefault()).toInstant())

But since you're just modeling a lastUpdated field, why not just `new Date())`?

ccho-mongodb replied:
Thanks, haven't kept up with Java 8+ dates. I'll change the field name and purpose since I want to show an example of setting a specific date.

jyemin commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,379 @@ Documents
+   * - Int64
+     - ``org.bson.BsonInt64``
+   * - Null
+     - ``null``
```
`null` is actually not allowed.  Use the `org.bson.BsonNull` type

ccho-mongodb replied:
Thanks, didn't see that type. Will update.

jyemin commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,379 @@ Documents
+   * - String
+     - ``org.bson.BsonString``
+
+The mapping table above shows the default mapping when working with the
```
You should not really customize the mapping for BsonDocument.  Let's remove this paragraph.

ccho-mongodb replied:
Will remove.

jyemin commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,379 @@ Documents
+   BsonDocument author = new BsonDocument()
+       .append("_id", new BsonObjectId())
+       .append("name", new BsonString("Gabriel García Márquez"))
+       .append("lastUpdated", new BsonDateTime(new Date(120, 2, 26).getTime()))
```
As above: just use `new Date()`

ccho-mongodb replied:
Will update for new field.

jyemin commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,379 @@ Documents
+.. code-block:: none
+   :copyable: false
+
+   _id: 5fb5fad05f734e3794741a35, name: Gabriel García Márquez, lastUpdated: Thu Mar 26 00:00:00 EDT 2020
```
I'm surprised you got 2020 given the date you used above

ccho-mongodb replied:
Will update for new field.

jyemin commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,379 @@ Documents
+
+   BasicDBObject author = new BasicDBObject("_id", new ObjectId())
+       .append("name", "Gabriel García Márquez")
+       .append("lastUpdated", new Date(120, 2, 26))
```
As above

ccho-mongodb replied:
Will update for new field.

jyemin commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -14,34 +14,379 @@ Documents
+Summary
+-------
+
+In this guide, we covered the following topics on Java document classes:
```
`Java document classes` is kind of awkward.  It's Java driver classes, not Java classes.  And they're not document classes, they're classes used to model/represent BSON documents.

Can you rephrase?

ccho-mongodb replied:
Sure, found a couple other similar mentions in this PR, will update.

ccho-mongodb commented:
Ready for re-review @jyemin 
-Updated Date instantiation
-Updated date field name in the sample document and output
-Removed text about customizing codec for BsonDocument
-Updated null to BsonNull type
-Clarified Java types from standard library
-Removed mention of document classes
-Removed mention of document as base unit in data hierarchy

jyemin commented on source/fundamentals/data-formats/documents.txt:
```diff
@@ -15,15 +15,14 @@ Overview
 A MongoDB **document** is a data structure that contains key/value pairs in
-binary JSON (BSON) format. The document is the base unit of the MongoDB
-data hierarchy: each document belongs to a single collection
+binary JSON (BSON) format. Each document belongs to a single collection
```
I still think this is misleading since we use documents for queries, commands, etc.  Those don't belong to a collection.

ccho-mongodb replied:
Ok thanks, I'll update the sentence. My intent is to build a little more context beyond the first sentence without having the user follow the link. If you think that's extraneous or if there's some other essential info you'd prefer to include instead, let me know.

ccho-mongodb commented:
Addressed https://github.com/mongodb/docs-java/pull/39#discussion_r530652834 
Ready for re-review.
